rgw/account: Support backward compatibility for s3:PutAcls calls for users migrated to account.

kchheda3 opened the pull request:
Fixes: https://tracker.ceph.com/issues/73759. 

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

kchheda3 committed:
rgw/account: Support backward compatibility for s3:PutAcls calls for users migrated to account.

fixes tracker https://tracker.ceph.com/issues/73759

Signed-off-by: kchheda3 <kchheda3@bloomberg.net>

jmundack commented:
jenkins test docs

kchheda3 committed:
rgw/account: Update the docs to add note about supporting backward compatibility for s3:PutAcls calls for users migrated to account.

Signed-off-by: kchheda3 <kchheda3@bloomberg.net>

anthonyeleven reviewed (commented):
Docs lgtm

bluikko commented:
> jenkins test docs

If you are able to rebase onto `main` the RTD task should be fixed.

Edit: it's `main` of course, not `master`.

cbodley commented on src/rgw/rgw_op.cc:
```diff
@@ -6333,8 +6333,12 @@ void RGWPutACLs::execute(optional_yield y)
+  // the requester may change between their user id and account id.
   if (!existing_owner.empty() &&
-      existing_owner.id != new_policy.get_owner().id) {
+      existing_owner.id != new_policy.get_owner().id &&
```
> if a role belonging to a tenant is trying to change the acl, will it be allowed to do so?

a tenant role would have an empty account so not match any account ids. that role would not be allowed to change the owner

but yeah, it looks an account role could still match the assuming user's id which is definitely not what we want

pritha-srivastava replied:
Ok, thanks for the explanation Casey. All my questions are answered. No objection from my end for merging this PR.

cbodley replied:
it would be great if you can take that commit here, so we can qa/backport them together

cbodley committed:
rgw/auth: don't mix role ownership between accounts and tenants

is_owner_of() for account roles should not try to match against the
assuming user id

Signed-off-by: Casey Bodley <cbodley@redhat.com>

bluikko commented on doc/radosgw/s3/bucketops.rst:
```diff
@@ -288,6 +288,8 @@ Request Entities
 | ``Permission``            | String      | The permission given to the ``Grantee`` bucket.                                              |
 +---------------------------+-------------+----------------------------------------------------------------------------------------------+
 
+.. note:: For accounts users, the ``Owner`` and ``ID`` in ``AccessControlPolicy`` may be set to either the legacy user ID or the new account ID. Both are accepted for backward compatibility.
```
I would suggest to link here to the accounts documentation. Can also do this myself in some future PR. Something like:
```console
.. note:: For :ref:`accounts users <radosgw-account>`, the [...]
```

Also, if touching this again, I suggest wrapping text lines before column 80.

bluikko commented on doc/radosgw/s3/objectops.rst:
```diff
@@ -234,6 +234,7 @@ Request Entities
 | ``Permission``            | String      | The permission given to the ``Grantee`` object.                                              |
 +---------------------------+-------------+----------------------------------------------------------------------------------------------+
 
+.. note:: For accounts users, the ``Owner`` and ``ID`` in ``AccessControlPolicy`` may be set to either the legacy user ID or the new account ID. Both are accepted for backward compatibility.
```
And same here.

bluikko reviewed (commented):
Docs LGTM.  
A nit and small suggestions for both the notes in case you happen to change something else in this PR. Otherwise I can also do this in another PR.

kchheda3 committed:
qa/rgw: add PutACL backward compatibility test for account migration

Test that put-bucket-acl and put-object-acl work both before and after
migrating a user to an account. After migration, the bucket/object ACL
owner is still the old user id, but the requester authenticates as the
account id.

Signed-off-by: Krunal Chheda <kchheda3@bloomberg.net>

kchheda3 commented:
> > @cbodley, confusion here in the test PR you commented that [ceph/s3-tests#708](https://github.com/ceph/s3-tests/pull/708) this is for vstart we need to add tests under "qa/workunits/rgw/" so you mean we first need to add tests under `qa/workunits/rgw/`
> 
> @kchheda3 i pushed a test_account_migration shell script to #66594 that we can build on for this testing. it works for me locally, but i'll need to verify the endpoint logic works in teuthology

@cbodley added migration test case, please do check once.

anrao19 commented:
Execution complete, tracker approved by @ivancich. tracker detail: https://tracker.ceph.com/issues/76249
@kchheda3, If no further testing needed, pr can be merged
